oak_crypto/src/noise_nk/mod.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This was copied from Chromium's third_party/cloud_authenticator, which has
+//! compatible copyright and ownership (Apache 2.0, Google).
+
+#[allow(unused_imports)] // Macros only used in tests.
+#[macro_use]
+
+mod error;
+mod noise;
+mod rustcrypto;
+
+use alloc::vec::Vec;
+use core::result::Result;
+
+pub use crate::noise_nk::rustcrypto::{
+    aes_256_gcm_open_in_place, aes_256_gcm_seal_in_place, ecdsa_verify, hkdf_sha256,
+    p256_scalar_mult, rand_bytes, sha256, sha256_two_part, EcdsaKeyPair, P256Scalar,
+};
+use crate::noise_nk::{
+    error::Error,
+    noise::{HandshakeType, Noise, NONCE_LEN},
+};
+
+// This is assumed to be vastly larger than any connection will ever reach.
+const MAX_SEQUENCE: u32 = 1u32 << 24;
+
+/// The length of an uncompressed, X9.62 encoding of a P-256 point.
+pub const P256_X962_LENGTH: usize = 65;
+
+#[derive(Debug)]
+pub struct Crypter {
+    read_key: [u8; 32],
+    write_key: [u8; 32],
+    read_nonce: u32,
+    write_nonce: u32,
+}
+
+/// Utility for encrypting and decrypting traffic between the Noise endpoints.
+/// It is created by |respond| and configured with a key for each traffic
+/// direction.
+impl Crypter {
+    fn new(read_key: &[u8; 32], write_key: &[u8; 32]) -> Self {
+        Self { read_key: *read_key, write_key: *write_key, read_nonce: 0, write_nonce: 0 }
+    }
+
+    fn next_nonce(nonce: &mut u32) -> Result<[u8; NONCE_LEN], Error> {
+        if *nonce > MAX_SEQUENCE {
+            return Err(Error::DecryptFailed);
+        }
+        let mut ret = [0u8; NONCE_LEN];
+        ret[NONCE_LEN - 4..].copy_from_slice(nonce.to_be_bytes().as_slice());
+        *nonce += 1;
+        Ok(ret)
+    }
+
+    pub fn encrypt(&mut self, plaintext: &[u8]) -> Result<Vec<u8>, Error> {
+        const PADDING_GRANULARITY: usize = 32;
+        static_assertions::const_assert!(PADDING_GRANULARITY < 256);
+        static_assertions::const_assert!((PADDING_GRANULARITY & (PADDING_GRANULARITY - 1)) == 0);
+
+        let mut padded_size: usize = plaintext.len();
+        if padded_size > (1usize << 28) {
+            return Err(Error::DataTooLarge(padded_size));
+        }
+        padded_size += 1; // padding-length byte
+        padded_size = (padded_size + PADDING_GRANULARITY - 1) & !(PADDING_GRANULARITY - 1);
+
+        let mut padded_encrypt_data = Vec::with_capacity(padded_size);
```

Comment:
Perhaps, but it also probably causes the generated code to initialize the plaintext destination with 0's rather than waiting for plaintext to overwrite the uninitialized area.  In low-level crypto, speed counts.